### ISSUE TITLE
Undocumented automatic fallback behaviour if multiple dep names are specified

### ISSUE THREAD

**Issue by Tachi107** [mesonbuild/meson#10715]
#### Describe the bug

The documentation doesn't state what happens when `dependency()` is called with multiple dependency names, and none of them are found. In particular, it doesn't say explicitly what subprojects are checked when the automatic fallback is enabled.

#### To Reproduce

    $ cat meson.build
    project('multiple_dep_names_fallback', 'c')
    dependency('first', 'second', 'third')

    $ cat subprojects/first/meson.build 
    project('first')
    meson.override_dependency('first', declare_dependency(compile_args: '-Dfirst'))

    $ cat subprojects/second/meson.build
    project('second')
    meson.override_dependency('second', declare_dependency(compile_args: '-Dsecond'))

    $ cat subprojects/third/meson.build 
    project('third')
    meson.override_dependency('third', declare_dependency(compile_args: '-Dthird'))

#### Current behavior

-   If the above code is executed as it is, the `first` subproject overrides the dependency and everything works as expected.

        Looking for a fallback subproject for the dependency first
        
        Executing subproject first 
        
        first| Project name: first
        first| Project version: undefined
        first| Build targets in project: 0
        first| Subproject first finished.
        
        Dependency first found: YES undefined (overridden)
        Build targets in project: 0
        
        multiple_dep_names_fallback undefined
        
          Subprojects
            first: YES

-   If the `first` subproject is removed, the `second` subproject overrides the dependency. Again, nothing strange going on.

        Looking for a fallback subproject for the dependency first
        
        Executing subproject second 
        
        second| Project name: second
        second| Project version: undefined
        second| Build targets in project: 0
        second| Subproject second finished.
        
        Dependency first found: YES undefined (overridden)
        Build targets in project: 0
        
        multiple_dep_names_fallback undefined
        
          Subprojects
            second: YES

-   If the `second` subproject is altered to override a dependency named `third`, the dependency is overridden again, and is satisfied. This kinda makes sense, as the subproject is overriding one of the searched dependencies.

        Looking for a fallback subproject for the dependency first
        
        Executing subproject second 
        
        second| Project name: second
        second| Project version: undefined
        second| Build targets in project: 0
        second| Subproject second finished.
        
        Dependency first found: YES undefined (overridden)
        Build targets in project: 0
        
        multiple_dep_names_fallback undefined
        
          Subprojects
            second: YES

-   If the `second` subproject is altered again to override a dependency named `unrelated`, dependency resolution fails, and the `third` subproject isn't even checked. I would've expected Meson to also look for the `third` subproject, because it did not fail when `first` wasn't found, so why should it fail if `second` doesn't override the searched dependency? Also, the error message is inaccurate, as the `second` subproject could've overridden a dependency named either `first`, `second`, or `third`.

        Looking for a fallback subproject for the dependency first
        
        Executing subproject second 
        
        second| Project name: second
        second| Project version: undefined
        second| Build targets in project: 0
        second| Subproject second finished.
        
        WARNING: Subproject 'second' did not override 'first' dependency and no variable name specified
        Dependency first from subproject subprojects/second found: NO
        
        meson.build:2:0: ERROR: Dependency 'first' is required but not found.

I think it might be appropriate to document and/or address the current behaviour, while also adjusting the error messages. This could also make things easier for muon (Cc: @annacrombie)

#### System parameters

* Python 3.10.6
* Meson 0.62.2

**Comment by xclaesse** [comment]
> * If the `second` subproject is altered again to override a dependency named `unrelated`, dependency resolution fails, and the `third` subproject isn't even checked.

I think that's the expected behaviour, if your subproject claims to provide dependency named "second", it must override it, otherwise that's an error. This is not supposed to be used for totally unrelated names, usually that would be stuff like "foo", "foo-2.0", "FOO", when the same library can be named sligthly differently on different systems. Basically Meson treat all those names as aliases.

**Comment by Tachi107** [comment]
> I think that's the expected behaviour, if your subproject claims to provide dependency named "second", it must override it, otherwise that's an error.

Yeah, that's indeed an error when using a single fallback, but in this case Meson has three different options, and aborting the build because one of the available fallbacks didn't provide the searched dependency doesn't seem right, especially because looking for the `third` subproject would've satisfied the dependency.

Also, not overriding the dependency is a warning, as you can see by the message. The error is triggered outside of the subproject scope, as Meson declares that the required dependency was not found.

Lastly, in this case the `second` subproject never declared to provide a dependency named `second`, as a project named `second` doesn't necessarily have to override a dependency with the same name. Otherwise the third example wouldn't have worked.

**Comment by annacrombie** [comment]
muon's current behavior is consistent with @Tachi107's expectations (and inconsistent with meson).  TBH dependency resolution with multiple fallbacks, subprojects, override_dependency, subproject promotion, etc. makes dependency() get very hard to reason about fast, unless you only do simple stuff (single dependency name with _maybe_ a fallback in subprojects/).  I think actually speccing out the behavior would present many opportunities for improvement.  I for one would prefer a less-magical dependency().

**Comment by eli-schwartz** [comment]
> Lastly, in this case the second subproject never declared to provide a dependency named second, as a project named second doesn't necessarily have to override a dependency with the same name. Otherwise the third example wouldn't have worked.

This is a factual inaccuracy on your part. If a subproject doesn't claim to provide a dependency of some kind of name, and it isn't explicitly used as `subproject()`, then it won't be used in the first place by dependency fallback.

The thing is that there's several different ways to "claim to provide a dependency". The most explicit way to do it is via the `[provide]` section of a wrap file. The other way to do it is for meson to try to configure a subproject named "foo" and see if it provides an override_dependency() matching the same name string as the subproject name string (i.e. the directory name of the subproject).

If you do `dependency('second')` and meson tries to configure "second" to find "override_dependency('second')" and that fails, because it provides "third" instead... then that's rightly going to be a failed fallback.

The fact that you used `dependency('second', 'third')` doesn't mean Meson should do a dependency search for "second", find "third", and be okay with that. It actually makes sense for it to just treat that subproject as a failed attempt and discard that entire line of thought.

It's not a real world situation IMO, because people aren't going to change a subproject to return something that it isn't for something other than fuzzing the behavior.

But I do agree that this could be handled better, in two ways:
- we could document this better.
- if a dependency with multiple names tries a subproject and fails, it could be argued to treat it two different ways: a) that subproject is supposed to provide it, but fails, so the dependency fails because its lookup is broken, or b) resolution of "second" failed but we should still try the name "third" and look up subprojects/third/ to do it. I would be inclined to say we should follow option b, for the same reason that if cmake finds a package but it doesn't provide the right modules, I'm pretty sure we still fall back to subprojects.

Furthermore:

> the error message is inaccurate, as the second subproject could've overridden a dependency named either `first`, `second`, or `third`.

It may override whatever it wants, but as a dependency fallback it's only announced to override "second" and that's the only thing that Meson is looking for it to override.

The error message is of course incorrect to claim that it needs to override "first" -- dependencies with multiple names are pretty new and using error messages from before they were added. This should be fixed.

**Comment by Tachi107** [comment]
> The thing is that there's several different ways to "claim to provide a dependency". The most explicit way to do it is via the `[provide]` section of a wrap file. The other way to do it is for meson to try to configure a subproject named "foo" and see if it provides an override_dependency() matching the same name string as the subproject name string (i.e. the directory name of the subproject).

That's right, thanks for pointing it out.

> If you do `dependency('second')` and meson tries to configure "second" to find "override_dependency('second')" and that fails, because it provides "third" instead... then that's rightly going to be a failed fallback.
> 
> The fact that you used `dependency('second', 'third')` doesn't mean Meson should do a dependency search for "second", find "third", and be okay with that. It actually makes sense for it to just treat that subproject as a failed attempt and discard that entire line of thought.

Expect that Meson is completely OK with this, see the third example: subproject `second` overrides a dependency named `third` and dependency resolution succeeds.

Dependency resolution currently fails if one of the subprojects exists but does not override either of the dependency names specified in the `dependency()` call.

> - if a dependency with multiple names tries a subproject and fails, it could be argued to treat it two different ways: a) that subproject is supposed to provide it, but fails, so the dependency fails because its lookup is broken, or b) resolution of "second" failed but we should still try the name "third" and look up subprojects/third/ to do it. I would be inclined to say we should follow option b, for the same reason that if cmake finds a package but it doesn't provide the right modules, I'm pretty sure we still fall back to subprojects.

I'd too prefer behaviour b.

> > the error message is inaccurate, as the second subproject could've overridden a dependency named either `first`, `second`, or `third`.
>
> It may override whatever it wants, but as a dependency fallback it's only announced to override "second" and that's the only thing that Meson is looking for it to override.

That's not what's happening, see what I wrote a few lines above.

**Comment by eli-schwartz** [comment]
Yeah sorry I misread that and it actually seems weird. One way or another, we should spec out -- and test -- what this does, because it is not obvious.

**Comment by annacrombie** [comment]
Here is a rough-draft flowchart: https://gist.github.com/annacrombie/fcff5ea0fd5445b5aabeb9c4194dc385

**Comment by Tachi107** [comment]
I think that muon's behaviour seems reasonable, but the flow chart doesn't show what should happen in a situation like example number 3, where you are looking for `first`, `second`, and `third` and subproject `second` overrides dependency `third`. Meson currently considers this OK.

**Comment by annacrombie** [comment]
I've updated the wording to try and make it more clear, but both meson and muon check for if the subproject overrides _any_ of the names passed to dependency().  I think the difference here is that meson gives up once any subproject fallback is configured successfully, but doesn't override any dependency names.  

I think the reason that meson's behavior seems wrong is that if you have just one dependency name, and it turns out to be a subproject that doesn't override properly, you wind up with a not-found dependency (unless it is required of course).  The way I conceptualized dependency(name, name, name...) is that it is roughly equivalent to:
```
foreach name : names
    dep = dependency(name, required: false)
    if dep.found()
      break
    endif
endforeach
```
IOW, any case that causes `name` to be not-found, results in trying the next name, not giving up.
